feat(query): skip aggregate present for shard key regex aggregate query

TanviBhavsar opened the pull request:
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
 
Skip AggregatePresenter for individual queries generated for aggregate shard key regex query. Add AggregatePresenter after aggregating all individual queries. For average function AvgSampl would be used which will have value and count required to calculate final average. Individual queries would have average schema having timestamp, value and count

vishramachandran reviewed (changes requested):
Some very good changes here. Next time, you should split such big PRs into multiple to keep low-risk changes out to separate commit. For example the QueryContext refactor could have been a low-risk PR that would be easy to complete.

Anyway, good changes. Some questions and comments. If we need in-person discussion, lets have it.

TanviBhavsar commented on http/src/main/scala/filodb/http/PrometheusApiRoute.scala:
```diff
@@ -48,7 +48,7 @@ class PrometheusApiRoute(nodeCoord: ActorRef, settings: HttpSettings)(implicit a
           // No cross-cluster failure routing in this API, hence we pass empty config
           askQueryAndRespond(dataset, logicalPlan, explainOnly.getOrElse(false), verbose.getOrElse(false),
-            spread, PromQlQueryParams(query, start.toLong, step.toLong, end.toLong, spread), histMap.getOrElse(false))
+            spread, PromQlQueryParams(query, start.toLong, step.toLong, end.toLong), histMap.getOrElse(false))
```
I didn't add here because shardkey regex is not supported with FiloDb Http

TanviBhavsar commented on query/src/main/scala/filodb/query/exec/PromQlRemoteExec.scala:
```diff
@@ -99,9 +101,12 @@ case class PromQlRemoteExec(queryEndpoint: String,
     ColumnInfo("h", ColumnType.HistogramColumn))
   private val histRecSchema = SerializedRangeVector.toSchema(histColumns)
   private val histResultSchema = ResultSchema(histColumns, 1)
+  private val avgColumns = defaultColumns :+ ColumnInfo("count", ColumnType.LongColumn)
+  private val avgRecSchema = SerializedRangeVector.toSchema(avgColumns)
+  private val avgResultSchema = ResultSchema(avgColumns, 1)
```
Yes I am adding the columns and schemas to a map

TanviBhavsar commented:
> Some very good changes here. Next time, you should split such big PRs into multiple to keep low-risk changes out to separate commit. For example the QueryContext refactor could have been a low-risk PR that would be easy to complete.
> 
> Anyway, good changes. Some questions and comments. If we need in-person discussion, lets have it.
> 
> Please place in bold comments the big TODO on the table. We should not piggy-back on PromQL API to talk across clusters. We should come up with our own contract.

Thanks Vish! I have addressed your review comments

tjackpaul reviewed (approved):
lgtm.. thanks for this PR @TanviBhavsar .. 
i agree with Vish on splitting this PR into separate PRs one for queryContext cleanup and another one for skipAggregatePresent would help us in reverting/releasing the latter separately if required easily.. please consider that.. also, i hope the CLI is tested and not having any issues..

vishramachandran commented on query/src/test/scala/filodb/query/PromCirceSupportSpec.scala:
```diff
@@ -0,0 +1,98 @@
+                  |					[
+                  |						1601491649,
+                  |						"15.186417982460787",
+                  |						5
```
How will NaN be encoded ?

TanviBhavsar replied:
It will be a string "NaN". I have added it to test case

vishramachandran reviewed (approved):
Lets do some religious regression testing for this. Thanks!
